Enhanced timestamp format handling in AutoIntervalShardingAlgorithm

linghengqian opened the pull request:
Fixes #13069 .

Changes proposed in this pull request:
- Provides the `AutoIntervalShardingAlgorithm` with the ability to specify milliseconds in arbitrary digits.
- Add corresponding tests.
- Correspondingly modified Chinese and English documents.
- For `Math.ceil()`, specify that the type passed in is `double` instead of `long`. (The original division of two `long` variables is still `long`)

linghengqian commented on shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/AutoIntervalShardingAlgorithmTest.java:
```diff
@@ -119,4 +119,40 @@ public void assertRangeDoShardingWithGreaterTenTables() {
+        shardingAlgorithm.getProps().setProperty("sharding-seconds", "1");
+        shardingAlgorithm.init();
+        List<String> availableTargetNames = new LinkedList<>();
+        for (int i = 0; i < 32; i++) {
```
If needed, I can convert to `List<String> availableTargetNames = IntStream.range(0, 32).mapToObj(i -> "t_order_" + i).collect(Collectors.toCollection(LinkedList::new));`

totalo replied:
I think it's acceptable and doesn't need to be changed.

linghengqian commented:
I noticed that the problems with CI/CD are like `Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (validate) on project shardingsphere-sharding-core: You have 6 Checkstyle violations.`, Can anyone tell me what to do with Checkstyle? Or point me to a link.

totalo commented:
> I noticed that the problems with CI/CD are like `Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (validate) on project shardingsphere-sharding-core: You have 6 Checkstyle violations.`, Can anyone tell me what to do with Checkstyle? Or point me to a link.

You can swipe up to see the specific error.

linghengqian commented:
> > I noticed that the problems with CI/CD are like `Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (validate) on project shardingsphere-sharding-core: You have 6 Checkstyle violations.`, Can anyone tell me what to do with Checkstyle? Or point me to a link.
> 
> You can swipe up to see the specific error. such as: ![image](https://user-images.githubusercontent.com/29777558/151386575-971dd01e-4de9-4337-8165-42344781031a.png)
> 
> or you can execute `mvn clean compile` on your local.

Thanks for the reminder, I have resolved it.

sandynz commented:
Refer to original code and modified code in this PR, `DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")` is not changed, I did some date parsing test:
```
    private void pr15131ParseTest() {
        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        List<String> timeStrings = Arrays.asList("2021-11-09 20:35:03", "2021-11-09 20:35:03.6", "2021-11-09 20:35:03.67", "2021-11-09 20:35:03.678");
        for (String each : timeStrings) {
            LocalDateTime localDateTime = LocalDateTime.from(dateTimeFormatter.parse(each, new ParsePosition(0)));
            String formatted = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(localDateTime.toInstant(ZoneOffset.of("+8")).toEpochMilli());
            log.info("each={}, localDateTime={}, formatted={}", each, localDateTime, formatted);
        }
    }
```
Result:
```
each=2021-11-09 20:35:03, localDateTime=2021-11-09T20:35:03, formatted=2021-11-09 20:35:03.000
each=2021-11-09 20:35:03.6, localDateTime=2021-11-09T20:35:03, formatted=2021-11-09 20:35:03.000
each=2021-11-09 20:35:03.67, localDateTime=2021-11-09T20:35:03, formatted=2021-11-09 20:35:03.000
each=2021-11-09 20:35:03.678, localDateTime=2021-11-09T20:35:03, formatted=2021-11-09 20:35:03.000
```
Looks the milliseconds are erased.

Since `datetime-lower` and `datetime-upper` could specify milliseconds, I'm afraid it might cause issue. e.g.
```
datetime-lower=2020-01-01 00:00:00.456
datetime-upper=2020-01-01 00:00:00.789
```
For value `2020-01-01 00:00:00.123` and `2020-01-01 00:00:00.890`, which sharding will it be routed to?
I suggest to add related unit test to verify it. I'll mark it in review comment.

sandynz commented on shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/AutoIntervalShardingAlgorithm.java:
```diff
@@ -64,16 +65,16 @@
         dateTimeLower = getDateTime(DATE_TIME_LOWER_KEY);
         shardingSeconds = getShardingSeconds();
-        autoTablesAmount = (int) (Math.ceil(parseDate(props.getProperty(DATE_TIME_UPPER_KEY)) / shardingSeconds) + 2);
+        autoTablesAmount = (int) (Math.ceil((double) (parseDate(props.getProperty(DATE_TIME_UPPER_KEY)) / shardingSeconds)) + 2);
```
Question: `parseDate` return `long` (just as PR says), is `(double)` casting necessary?
And I found there's `(float)` casting in `doSharding` method, should it be `(double)` too?

linghengqian replied:
> Question: `parseDate` return `long` (just as PR says), is `(double)` casting necessary? And I found there's `(float)` casting in `doSharding` method, should it be `(double)` too?

For the change of the `Math.ceil()` parameter, because the input parameter of the `Math.ceil()` function is specified as `double`, so I do an explicit rather than an implicit division result of the `long` type The cast. This doesn't cause problems with the program, but I think explicit casts are good for IDE (this is reflected in `IntelliJ IDEA` as `Integer division in floating point context`) checks. I made synchronous changes , at the same time I did two tests, both with the same result.
````java
public void testFirst() {
DateTimeFormatter DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        LocalDateTime dateTimeLower = LocalDateTime.parse("2020-01-01 00:00:00", DATE_TIME_FORMAT);
        System.out.println("dateTimeLower is----------------->" + dateTimeLower);
        int autoTablesAmount = (int) (Math.ceil((double) (Duration.between(dateTimeLower,
                LocalDateTime.parse("2022-12-28 00:00:00", DATE_TIME_FORMAT)).toMillis() / 1000 / shardingSeconds)) + 2);
        System.out.println("autoTablesAmount is---------------------------->" + autoTablesAmount);
        LocalDateTime dateValue = LocalDateTime.parse("2021-10-11 00:00:00", DATE_TIME_FORMAT);
        System.out.println("dateValue is-------------------->" + dateValue);
        long shardingValue = Duration.between(dateTimeLower, dateValue).toMillis() / 1000;
        System.out.println("shardingValue is------------------->" + shardingValue);
        String position = new DecimalFormat("0.00").format((double) shardingValue / shardingSeconds);
        System.out.println("position is-------------------------->" + position);
        int tableNameSuffix = Math.min(Math.max(0, (int) Math.ceil(Double.parseDouble(position))), autoTablesAmount - 1);
        System.out.println("tableNameSuffix is--------------->" + tableNameSuffix);
    }
````
Another test is the algorithm test before submitting the PR.
````java
public void testSecond() {
        long shardingSeconds = 604800L;
        DateTimeFormatter DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        LocalDateTime dateTimeLower = LocalDateTime.parse("2020-01-01 00:00:00", DATE_TIME_FORMAT);
        System.out.println("dateTimeLower is----------------->" + dateTimeLower);
        int autoTablesAmount = (int) (Math.ceil(Duration.between(dateTimeLower,
                LocalDateTime.parse("2022-12-28 00:00:00", DATE_TIME_FORMAT)).toMillis() / 1000 / shardingSeconds) + 2);
        System.out.println("autoTablesAmount is---------------------------->" + autoTablesAmount);
        LocalDateTime dateValue = LocalDateTime.parse("2021-10-11 00:00:00", DATE_TIME_FORMAT);
        System.out.println("dateValue is-------------------->" + dateValue);
        long shardingValue = Duration.between(dateTimeLower, dateValue).toMillis() / 1000;
        System.out.println("shardingValue is------------------->" + shardingValue);
        String position = new DecimalFormat("0.00").format((float) shardingValue / shardingSeconds);
        System.out.println("position is-------------------------->" + position);
        int tableNameSuffix = Math.min(Math.max(0, (int) Math.ceil(Float.parseFloat(position))), autoTablesAmount - 1);
        System.out.println("tableNameSuffix is--------------->" + tableNameSuffix);
    }
````

sandynz replied:
That's OK

sandynz commented on shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/AutoIntervalShardingAlgorithmTest.java:
```diff
@@ -119,4 +119,45 @@ public void assertRangeDoShardingWithGreaterTenTables() {
+        }
+        Collection<String> actualWithoutMilliseconds = shardingAlgorithm.doSharding(availableTargetNames,
+                new RangeShardingValue<>("t_order", "create_time", Range.closed("2020-01-01 00:00:11", "2020-01-01 00:00:21")));
+        assertThat(actualWithoutMilliseconds.size(), is(11));
```
Could we add more test cases for `doSharding` to verify certain value's sharding target, but not only the target size? (I've added comment in PR about date parsing)

linghengqian replied:
> Could we add more test cases for `doSharding` to verify certain value's sharding target, but not only the target size? (I've added comment in PR about date parsing)

I added the corresponding instructions below the PR and removed this test.

sandynz replied:
That's OK

linghengqian commented:
@sandynz 
- I did originally plan to change `private static final DateTimeFormatter DATE_TIME_FORMAT`, the following may be what you want. `Java8 LocalTime` is different from `MySQL DateTime`, the time precision of `MySQL DateTime` supports up to 6 decimal places. is the reason for `.appendFraction(ChronoField.MILLI_OF_SECOND,0,6,true)` in the test case.
````java
private void pr15131ParseTest() {
       DateTimeFormatter dateTimeFormatter = new DateTimeFormatterBuilder()
                .appendPattern("yyyy-MM-dd HH:mm:ss")
                .appendFraction(ChronoField.MILLI_OF_SECOND,0,6,true)
                .toFormatter();
        List<String> timeStrings = Arrays.asList("2021-11-09 20:35:03", "2021-11-09 20:35:03.6", "2021-11-09 20:35:03.67", " 2021-11-09 20:35:03.678");
        for (String each : timeStrings) {
            LocalDateTime localDateTime = LocalDateTime.from(dateTimeFormatter.parse(each, new ParsePosition(0)));
            String formatted = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(localDateTime.toInstant(ZoneOffset.of("+8")).toEpochMilli());
            log.info("each={}, localDateTime={}, formatted={}", each, localDateTime, formatted);
        }
    }
````
turn out
````java
each=2021-11-09 20:35:03, localDateTime=2021-11-09T20:35:03, formatted=2021-11-09 20:35:03.000
each=2021-11-09 20:35:03.6, localDateTime=2021-11-09T20:35:03.600, formatted=2021-11-09 20:35:03.600
each=2021-11-09 20:35:03.67, localDateTime=2021-11-09T20:35:03.670, formatted=2021-11-09 20:35:03.670
each=2021-11-09 20:35:03.678, localDateTime=2021-11-09T20:35:03.678, formatted=2021-11-09 20:35:03.678
````
- For the test that needs to be added, I think I should explain, because the unit of `AutoIntervalShardingAlgorithm` is only `seconds`, it does not allow to modify this unit. And there is a step in the function `Duration.between(dateTimeLower, dateValueFirst).toMillis() / 1000;` , this will automatically erase all time precision after the second. I've provided my test within this `pull request`, proving that this step will cause something like `2021-10-11 00:00:00.123` Time maps to `2021-10-11 00:00:00`.
````java
 public static void main(String[] args) {
        DateTimeFormatter DATE_TIME_FORMAT = new DateTimeFormatterBuilder()
                .appendPattern("yyyy-MM-dd HH:mm:ss")
                .appendFraction(ChronoField.MILLI_OF_SECOND,0,3,true)
                .toFormatter();
        LocalDateTime dateTimeLower = LocalDateTime.parse("2020-01-01 00:00:00", DATE_TIME_FORMAT);
        LocalDateTime dateValueFirst = LocalDateTime.parse("2021-10-11 00:00:00", DATE_TIME_FORMAT);
        LocalDateTime dateValueSecond = LocalDateTime.parse("2021-10-11 00:00:00.123", DATE_TIME_FORMAT);
        long shardingValueFirst = Duration.between(dateTimeLower, dateValueFirst).toMillis() / 1000;
        System.out.println("shardingValueFirst-->"+shardingValueFirst);
        long shardingValueSecond = Duration.between(dateTimeLower, dateValueFirst).toMillis() / 1000;
        System.out.println("shardingValueSecond-->"+shardingValueSecond);
    }
````
- I will not propose changes like `Duration.between(dateTimeLower, dateValueFirst).toNanos() / 1000;` in this PR, because in my practice this change is completely different from the original intention of this class, Maybe a new algorithm class is needed to deal with nanoseconds.
- I just want to make the read value with milliseconds usable by the algorithm class instead of throwing an exception `could not be parsed`. So in the new commit I removed `private` for `Property` A change to the `private LocalDateTime getDateTime()` function, which is an attempt of mine in the past, and it turns out that this change is not necessary. I changed the description of the documentation synchronously.
- I submit a new `commit` for such issues.
